File: FloatItPhone/src/main/java/net/geekstools/floatshort/PRO/Folders/FoldersAdapter/AppSelectionListAdapter.java
/*
 * Copyright © 2020 By Geeks Empire.
 *
 * Created by Elias Fazel on 3/26/20 2:51 PM
 * Last modified 3/26/20 2:26 PM
 *
 * Licensed Under MIT License.
 * https://opensource.org/licenses/MIT
 */

package net.geekstools.floatshort.PRO.Folders.FoldersAdapter;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.content.res.ColorStateList;
import android.util.DisplayMetrics;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.TranslateAnimation;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import net.geekstools.floatshort.PRO.Folders.Utils.ConfirmButtonProcessInterface;
import net.geekstools.floatshort.PRO.R;
import net.geekstools.floatshort.PRO.Utils.AdapterItemsData.AdapterItems;
import net.geekstools.floatshort.PRO.Utils.Functions.FunctionsClass;
import net.geekstools.floatshort.PRO.Utils.Functions.FunctionsClassSecurity;
import net.geekstools.floatshort.PRO.Utils.Functions.PublicVariable;
import net.geekstools.floatshort.PRO.Utils.UI.CustomIconManager.LoadCustomIcons;
import net.geekstools.floatshort.PRO.databinding.AdvanceAppSelectionListBinding;
import net.geekstools.imageview.customshapes.ShapesImage;

import java.io.File;
import java.util.ArrayList;

public class AppSelectionListAdapter extends RecyclerView.Adapter<AppSelectionListAdapter.ViewHolder> {

    private Context context;

    FunctionsClass functionsClass;
    FunctionsClassSecurity functionsClassSecurity;

    ConfirmButtonProcessInterface confirmButtonProcessInterface;

    ImageView temporaryFallingIcon;

    float fromX, fromY, toX, toY, dpHeight, dpWidth, systemUiHeight;
    int animationType, layoutInflater;

    LoadCustomIcons loadCustomIcons;

    private ArrayList<AdapterItems> adapterItems;

    public AppSelectionListAdapter(Context context, AdvanceAppSelectionListBinding advanceAppSelectionListBinding, 
                                   ArrayList<AdapterItems> adapterItems,
                                   ConfirmButtonProcessInterface confirmButtonProcessInterface) {
        this.context = context;
        this.adapterItems = adapterItems;

        this.confirmButtonProcessInterface = confirmButtonProcessInterface;

        functionsClass = new FunctionsClass(context);
        functionsClassSecurity = new FunctionsClassSecurity(context);


        temporaryFallingIcon = functionsClass.initShapesImage(advanceAppSelectionListBinding.temporaryFallingIcon);

        DisplayMetrics displayMetrics = context.getResources().getDisplayMetrics();
        dpHeight = displayMetrics.heightPixels;
        dpWidth = displayMetrics.widthPixels;
        systemUiHeight = PublicVariable.actionBarHeight;
        fromX = toX = PublicVariable.confirmButtonX;
        toY = PublicVariable.confirmButtonY;
        animationType = Animation.ABSOLUTE;

        switch (functionsClass.shapesImageId()) {
            case 1:
                layoutInflater = R.layout.selection_item_card_list_droplet;
                break;
            case 2:
                layoutInflater = R.layout.selection_item_card_list_circle;
                break;
            case 3:
                layoutInflater = R.layout.selection_item_card_list_square;
                break;
            case 4:
                layoutInflater = R.layout.selection_item_card_list_squircle;
                break;
            case 0:
                layoutInflater = R.layout.selection_item_card_list_noshape;
                break;
        }

        if (functionsClass.customIconsEnable()) {
            loadCustomIcons = new LoadCustomIcons(context, functionsClass.customIconPackageName());
        }
    }

    @Override
    public AppSelectionListAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

        return new ViewHolder(LayoutInflater.from(context).inflate(layoutInflater, parent, false));
    }

    @SuppressLint("ClickableViewAccessibility")
    @Override
    public void onBindViewHolder(ViewHolder viewHolderBinder, final int position) {
        try {
            final String packageName = adapterItems.get(position).getPackageName();
            File autoFile = context.getFileStreamPath(packageName + PublicVariable.categoryName);
            viewHolderBinder.autoChoice.setChecked(false);
            if (autoFile.exists()) {
                viewHolderBinder.autoChoice.setChecked(true);
            } else {
                viewHolderBinder.autoChoice.setChecked(false);
            }

            if (PublicVariable.themeLightDark) {
                viewHolderBinder.autoChoice.setButtonTintList(ColorStateList.valueOf(context.getColor(R.color.dark)));
            } else if (!PublicVariable.themeLightDark) {
                viewHolderBinder.appName.setTextColor(context.getColor(R.color.light));
                viewHolderBinder.autoChoice.setButtonTintList(ColorStateList.valueOf(context.getColor(R.color.light)));
            }

            viewHolderBinder.appIcon.setImageDrawable(adapterItems.get(position).getAppIcon());
            viewHolderBinder.appName.setText(adapterItems.get(position).getAppName());
        } catch (Exception e) {
            e.printStackTrace();
        }

        viewHolderBinder.item.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View view, MotionEvent motionEvent) {
                switch (motionEvent.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        fromY = -((dpHeight - motionEvent.getRawY()) - (systemUiHeight));
                        break;
                    case MotionEvent.ACTION_UP:
                        final String pack = adapterItems.get(position).getPackageName();
                        File autoFile = context.getFileStreamPath(pack + PublicVariable.categoryName);
                        if (autoFile.exists()) {
                            context.deleteFile(pack + PublicVariable.categoryName);
                            functionsClass.removeLine(PublicVariable.categoryName, adapterItems.get(position).getPackageName());
                            viewHolderBinder.autoChoice.setChecked(false);

                            confirmButtonProcessInterface.savedShortcutCounter();

                            confirmButtonProcessInterface.hideSavedShortcutList();
                            context.sendBroadcast(new Intent(context.getString(R.string.visibilityActionAdvance)));

                            if (functionsClassSecurity.isAppLocked(PublicVariable.categoryName)) {
                                functionsClassSecurity.doUnlockApps(adapterItems.get(position).getPackageName());
                            }
                        } else {
                            functionsClass.saveFile(
                                    pack + PublicVariable.categoryName, pack);
                            functionsClass.saveFileAppendLine(
                                    PublicVariable.categoryName, pack);
                            viewHolderBinder.autoChoice.setChecked(true);

                            if (functionsClassSecurity.isAppLocked(PublicVariable.categoryName)) {
                                functionsClassSecurity.doLockApps(adapterItems.get(position).getPackageName());
                            }

                            TranslateAnimation translateAnimation =
                                    new TranslateAnimation(animationType, fromX,
                                            animationType, toX,
                                            animationType, fromY,
                                            animationType, toY);
                            translateAnimation.setDuration((long) Math.abs(fromY));

                            temporaryFallingIcon.setImageDrawable(functionsClass.customIconsEnable() ?
                                    loadCustomIcons.getDrawableIconForPackage(adapterItems.get(position).getPackageName(), functionsClass.shapedAppIcon(adapterItems.get(position).getPackageName()))
                                    :
                                    functionsClass.shapedAppIcon(adapterItems.get(position).getPackageName()));
                            temporaryFallingIcon.startAnimation(translateAnimation);
                            translateAnimation.setAnimationListener(new Animation.AnimationListener() {
                                @Override
                                public void onAnimationStart(Animation animation) {
                                    confirmButtonProcessInterface.hideSavedShortcutList();
                                    context.sendBroadcast(new Intent(context.getString(R.string.visibilityActionAdvance)));

                                    temporaryFallingIcon.setVisibility(View.VISIBLE);
                                }

                                @Override
                                public void onAnimationEnd(Animation animation) {
                                    temporaryFallingIcon.setVisibility(View.INVISIBLE);

                                    context.sendBroadcast(new Intent(context.getString(R.string.animtaionActionAdvance)));
                                    confirmButtonProcessInterface.savedShortcutCounter();
                                }

                                @Override
                                public void onAnimationRepeat(Animation animation) {
                                }
                            });
                        }
                        break;
                }
                return true;
            }
        });
    }

    @Override
    public int getItemCount() {
        return adapterItems.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {

        RelativeLayout item;
        ShapesImage appIcon;
        TextView appName;
        CheckBox autoChoice;

        public ViewHolder(View view) {
            super(view);
            item = (RelativeLayout) view.findViewById(R.id.item);
            appIcon = (ShapesImage) view.findViewById(R.id.icon);
            appName = (TextView) view.findViewById(R.id.loadingDescription);
            autoChoice = (CheckBox) view.findViewById(R.id.autoChoice);
        }
    }
}

File: FloatItPhone/src/main/java/net/geekstools/floatshort/PRO/Folders/FoldersApplicationsSelectionProcess/Adapter/AppSelectionListAdapter.kt
/*
 * Copyright © 2020 By Geeks Empire.
 *
 * Created by Elias Fazel
 * Last modified 8/29/20 3:58 AM
 *
 * Licensed Under MIT License.
 * https://opensource.org/licenses/MIT
 */
package net.geekstools.floatshort.PRO.Folders.FoldersApplicationsSelectionProcess.Adapter

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.ColorStateList
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.TranslateAnimation
import android.widget.CheckBox
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import net.geekstools.floatshort.PRO.Folders.UI.AppsConfirmButton
import net.geekstools.floatshort.PRO.Folders.Utils.ConfirmButtonProcessInterface
import net.geekstools.floatshort.PRO.R
import net.geekstools.floatshort.PRO.SecurityServices.AuthenticationProcess.Utils.SecurityFunctions
import net.geekstools.floatshort.PRO.Utils.AdapterItemsData.AdapterItems
import net.geekstools.floatshort.PRO.Utils.Functions.FileIO
import net.geekstools.floatshort.PRO.Utils.Functions.FunctionsClassLegacy
import net.geekstools.floatshort.PRO.Utils.Functions.PublicVariable
import net.geekstools.floatshort.PRO.Utils.UI.CustomIconManager.LoadCustomIcons
import net.geekstools.floatshort.PRO.databinding.AdvanceAppSelectionListBinding
import net.geekstools.imageview.customshapes.ShapesImage
import java.util.*
import kotlin.math.abs

class AppSelectionListAdapter(private val context: Context,
                              private val functionsClassLegacy: FunctionsClassLegacy,
                              private val advanceAppSelectionListBinding: AdvanceAppSelectionListBinding,
                              private val adapterItems: ArrayList<AdapterItems>,
                              private val appsConfirmButton: AppsConfirmButton,
                              private val confirmButtonProcessInterface: ConfirmButtonProcessInterface) : RecyclerView.Adapter<AppSelectionListAdapter.ViewHolder>() {

    private val securityFunctions: SecurityFunctions = SecurityFunctions(context)

    private val fileIO: FileIO = FileIO(context)

    var temporaryFallingIcon: ImageView = functionsClassLegacy.initShapesImage(advanceAppSelectionListBinding.temporaryFallingIcon)

    object PickedAttribute {
        var fromX: Float = 0f
        var fromY: Float = 0f

        var toX: Float = 0f
        var toY: Float = 0f

        var dpHeight: Float = 0f
        var dpWidth: Float = 0f

        var systemUiHeight: Float = 0f

        var animationType: Int = 0
    }

    var layoutInflaterView = 0

    lateinit var loadCustomIcons: LoadCustomIcons

    init {
        val displayMetrics = context.resources.displayMetrics

        PickedAttribute.dpHeight = displayMetrics.heightPixels.toFloat()
        PickedAttribute.dpWidth = displayMetrics.widthPixels.toFloat()

        PickedAttribute.systemUiHeight = PublicVariable.actionBarHeight.toFloat()

        PickedAttribute.toX = PublicVariable.confirmButtonX
        PickedAttribute.fromX = PickedAttribute.toX

        PickedAttribute.toY = PublicVariable.confirmButtonY

        PickedAttribute.animationType = Animation.ABSOLUTE

        when (functionsClassLegacy.shapesImageId()) {
            1 -> layoutInflaterView = R.layout.selection_item_card_list_droplet
            2 -> layoutInflaterView = R.layout.selection_item_card_list_circle
            3 -> layoutInflaterView = R.layout.selection_item_card_list_square
            4 -> layoutInflaterView = R.layout.selection_item_card_list_squircle
            0 -> layoutInflaterView = R.layout.selection_item_card_list_noshape
        }

        if (functionsClassLegacy.customIconsEnable()) {
            loadCustomIcons = LoadCustomIcons(context, functionsClassLegacy.customIconPackageName())
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {

        return ViewHolder(LayoutInflater.from(context).inflate(layoutInflaterView, parent, false))
    }

    override fun getItemCount(): Int {

        return adapterItems.size
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onBindViewHolder(viewHolderBinder: ViewHolder, position: Int) {

        val pickedPackageNameList = adapterItems[position].packageName
        val pickedPackageNameFileList = context.getFileStreamPath(pickedPackageNameList + PublicVariable.folderName)

        viewHolderBinder.checkboxSelectItem.isChecked = false
        viewHolderBinder.checkboxSelectItem.isChecked = pickedPackageNameFileList.exists()

        if (PublicVariable.themeLightDark) {
            viewHolderBinder.checkboxSelectItem.buttonTintList = ColorStateList.valueOf(context.getColor(R.color.dark))
        } else if (!PublicVariable.themeLightDark) {
            viewHolderBinder.appName.setTextColor(context.getColor(R.color.light))
            viewHolderBinder.checkboxSelectItem.buttonTintList = ColorStateList.valueOf(context.getColor(R.color.light))
        }

        viewHolderBinder.appIcon.setImageDrawable(adapterItems[position].appIcon)
        viewHolderBinder.appName.text = adapterItems[position].appName

        viewHolderBinder.item.setOnTouchListener { view, motionEvent ->

            when (motionEvent.action) {
                MotionEvent.ACTION_DOWN -> {
                    PickedAttribute.fromY = -(PickedAttribute.dpHeight - motionEvent.rawY - PickedAttribute.systemUiHeight)
                }
                MotionEvent.ACTION_UP -> {
                    val pickedPackageName = adapterItems[position].packageName
                    val pickedPackageNameFile = context.getFileStreamPath(pickedPackageName + PublicVariable.folderName)

                    if (pickedPackageNameFile.exists()) {
                        context.deleteFile(pickedPackageName + PublicVariable.folderName)
                        fileIO.removeLine(PublicVariable.folderName, adapterItems[position].packageName)

                        viewHolderBinder.checkboxSelectItem.isChecked = false

                        confirmButtonProcessInterface.savedShortcutCounter()
                        confirmButtonProcessInterface.hideSavedShortcutList()
                        appsConfirmButton.makeItVisible()

                        if (securityFunctions.isAppLocked(PublicVariable.folderName)) {
                            securityFunctions.doUnlockApps(adapterItems[position].packageName)
                        }

                    } else {
                        fileIO.saveFile(pickedPackageName + PublicVariable.folderName, pickedPackageName)
                        fileIO.saveFileAppendLine(PublicVariable.folderName, pickedPackageName)

                        viewHolderBinder.checkboxSelectItem.isChecked = true
                        if (securityFunctions.isAppLocked(PublicVariable.folderName)) {
                            securityFunctions.doLockApps(adapterItems[position].packageName)
                        }

                        val translateAnimation = TranslateAnimation(PickedAttribute.animationType, PickedAttribute.fromX,
                                PickedAttribute.animationType, PickedAttribute.toX,
                                PickedAttribute.animationType, PickedAttribute.fromY,
                                PickedAttribute.animationType, PickedAttribute.toY)

                        translateAnimation.duration = abs(PickedAttribute.fromY).toLong()

                        temporaryFallingIcon.setImageDrawable(if (functionsClassLegacy.customIconsEnable()) {
                            loadCustomIcons.getDrawableIconForPackage(adapterItems[position].packageName, functionsClassLegacy.shapedAppIcon(adapterItems[position].packageName))
                        } else {
                            functionsClassLegacy.shapedAppIcon(adapterItems[position].packageName)
                        })
                        temporaryFallingIcon.startAnimation(translateAnimation)

                        translateAnimation.setAnimationListener(object : Animation.AnimationListener {
                            override fun onAnimationStart(animation: Animation) {
                                confirmButtonProcessInterface.hideSavedShortcutList()
                                appsConfirmButton.makeItVisible()
                                temporaryFallingIcon.visibility = View.VISIBLE
                            }

                            override fun onAnimationEnd(animation: Animation) {
                                temporaryFallingIcon.visibility = View.INVISIBLE
                                
                                appsConfirmButton.startCustomAnimation(null)
                                confirmButtonProcessInterface.savedShortcutCounter()
                            }

                            override fun onAnimationRepeat(animation: Animation) {}
                        })
                    }
                }
            }
            true
        }
    }

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var item: RelativeLayout = view.findViewById<View>(R.id.fullItemView) as RelativeLayout
        var appIcon: ShapesImage = view.findViewById<View>(R.id.appIconView) as ShapesImage
        var appName: TextView = view.findViewById<View>(R.id.appNameView) as TextView
        var checkboxSelectItem: CheckBox = view.findViewById<View>(R.id.checkboxSelectItem) as CheckBox
    }
}